From 9fe94edecac90c3740a9ee40445ed4e8dda9f265 Mon Sep 17 00:00:00 2001
From: James Morse <james.morse@arm.com>
Date: Mon, 22 Jan 2024 14:58:19 +0800
Subject: [PATCH 667/960] downstream: untested: arm_mpam: resctrl: Allow
 monitors to be configured

MPAM MSCs may have support for filtering reads or writes when monitoring
traffic. Resctrl has a configuration bitmap for which kind of accesses
should be monitored.

Bridge the gap where possible. MPAM only has a read/write bit, so not
all the combinations can be supported.

Signed-off-by: James Morse <james.morse@arm.com>
Signed-off-by: Zeng Heng <zengheng4@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 drivers/platform/mpam/mpam_devices.c  | 26 +++++++++++
 drivers/platform/mpam/mpam_internal.h |  9 ++++
 drivers/platform/mpam/mpam_resctrl.c  | 62 +++++++++++++++++++++++++++
 3 files changed, 97 insertions(+)

diff --git a/drivers/platform/mpam/mpam_devices.c b/drivers/platform/mpam/mpam_devices.c
index 79ab28d63e7f..2026d96f3adb 100644
--- a/drivers/platform/mpam/mpam_devices.c
+++ b/drivers/platform/mpam/mpam_devices.c
@@ -1074,6 +1074,32 @@ int mpam_msmon_read(struct mpam_component *comp, struct mon_cfg *ctx,
 	return err;
 }
 
+void mpam_msmon_reset_all_mbwu(struct mpam_component *comp)
+{
+	int idx, i;
+	unsigned long flags;
+	struct mpam_msc *msc;
+	struct mpam_msc_ris *ris;
+
+	if (!mpam_is_enabled())
+		return;
+
+	idx = srcu_read_lock(&mpam_srcu);
+	list_for_each_entry_rcu(ris, &comp->ris, comp_list) {
+		if (!mpam_has_feature(mpam_feat_msmon_mbwu, &ris->props))
+			continue;
+
+		msc = ris->msc;
+		spin_lock_irqsave(&msc->mon_sel_lock, flags);
+		for (i = 0; i < ris->props.num_mbwu_mon; i++) {
+			ris->mbwu_state[i].correction = 0;
+			ris->mbwu_state[i].reset_on_next_read = true;
+		}
+		spin_unlock_irqrestore(&msc->mon_sel_lock, flags);
+	}
+	srcu_read_unlock(&mpam_srcu, idx);
+}
+
 void mpam_msmon_reset_mbwu(struct mpam_component *comp, struct mon_cfg *ctx)
 {
 	int idx;
diff --git a/drivers/platform/mpam/mpam_internal.h b/drivers/platform/mpam/mpam_internal.h
index fac93795c318..dfcb0155a842 100644
--- a/drivers/platform/mpam/mpam_internal.h
+++ b/drivers/platform/mpam/mpam_internal.h
@@ -20,6 +20,12 @@ DECLARE_STATIC_KEY_FALSE(mpam_enabled);
 /* Value to indicate the allocated monitor is derived from the RMID index. */
 #define USE_RMID_IDX	(U16_MAX + 1)
 
+/*
+ * Only these event configuration bits are supported. MPAM can't know if
+ * data is being written back, these will show up as a write.
+ */
+#define MPAM_RESTRL_EVT_CONFIG_VALID	(READS_TO_LOCAL_MEM | NON_TEMP_WRITE_TO_LOCAL_MEM)
+
 static inline bool mpam_is_enabled(void)
 {
 	return static_branch_likely(&mpam_enabled);
@@ -236,6 +242,8 @@ struct mpam_msc_ris {
 struct mpam_resctrl_dom {
 	struct mpam_component	*comp;
 	struct rdt_domain	resctrl_dom;
+
+	u32			mbm_local_evt_cfg;
 };
 
 struct mpam_resctrl_res {
@@ -295,6 +303,7 @@ int mpam_apply_config(struct mpam_component *comp, u16 partid,
 int mpam_msmon_read(struct mpam_component *comp, struct mon_cfg *ctx,
 		    enum mpam_device_features, u64 *val);
 void mpam_msmon_reset_mbwu(struct mpam_component *comp, struct mon_cfg *ctx);
+void mpam_msmon_reset_all_mbwu(struct mpam_component *comp);
 
 int mpam_resctrl_online_cpu(unsigned int cpu);
 int mpam_resctrl_offline_cpu(unsigned int cpu);
diff --git a/drivers/platform/mpam/mpam_resctrl.c b/drivers/platform/mpam/mpam_resctrl.c
index 4eef4b73db87..c31f739513ad 100644
--- a/drivers/platform/mpam/mpam_resctrl.c
+++ b/drivers/platform/mpam/mpam_resctrl.c
@@ -304,6 +304,18 @@ void resctrl_arch_mon_ctx_free(struct rdt_resource *r, int evtid,
 	}
 }
 
+static enum mon_filter_options resctrl_evt_config_to_mpam(u32 local_evt_cfg)
+{
+	switch (local_evt_cfg) {
+	case READS_TO_LOCAL_MEM:
+		return COUNT_READ;
+	case NON_TEMP_WRITE_TO_LOCAL_MEM:
+		return COUNT_WRITE;
+	default:
+		return COUNT_BOTH;
+	}
+}
+
 int resctrl_arch_rmid_read(struct rdt_resource	*r, struct rdt_domain *d,
 			   u32 closid, u32 rmid, enum resctrl_event_id eventid,
 			   u64 *val, void *arch_mon_ctx)
@@ -337,6 +349,7 @@ int resctrl_arch_rmid_read(struct rdt_resource	*r, struct rdt_domain *d,
 
 	cfg.match_pmg = true;
 	cfg.pmg = rmid;
+	cfg.opts = resctrl_evt_config_to_mpam(dom->mbm_local_evt_cfg);
 
 	if (cdp_enabled) {
 		cfg.partid = closid << 1;
@@ -619,6 +632,54 @@ static void mpam_resctrl_pick_mba(void)
 	}
 }
 
+bool resctrl_arch_is_evt_configurable(enum resctrl_event_id evt)
+{
+	struct mpam_props *cprops;
+
+	switch (evt) {
+	case QOS_L3_MBM_LOCAL_EVENT_ID:
+		if (!mbm_local_class)
+			return false;
+		cprops = &mbm_local_class->props;
+
+		return mpam_has_feature(mpam_feat_msmon_mbwu_rwbw, cprops);
+	default:
+		return false;
+	}
+}
+
+void resctrl_arch_mon_event_config_read(void *info)
+{
+	struct mpam_resctrl_dom *dom;
+	struct resctrl_mon_config_info *mon_info = info;
+
+	dom = container_of(mon_info->d, struct mpam_resctrl_dom, resctrl_dom);
+	mon_info->mon_config = dom->mbm_local_evt_cfg & MAX_EVT_CONFIG_BITS;
+}
+
+void resctrl_arch_mon_event_config_write(void *info)
+{
+	struct mpam_resctrl_dom *dom;
+	struct resctrl_mon_config_info *mon_info = info;
+
+	if (mon_info->mon_config & ~MPAM_RESTRL_EVT_CONFIG_VALID) {
+		mon_info->err = -EOPNOTSUPP;
+		return;
+	}
+
+	dom = container_of(mon_info->d, struct mpam_resctrl_dom, resctrl_dom);
+	dom->mbm_local_evt_cfg = mon_info->mon_config & MPAM_RESTRL_EVT_CONFIG_VALID;
+}
+
+void resctrl_arch_reset_rmid_all(struct rdt_resource *r, struct rdt_domain *d)
+{
+	struct mpam_resctrl_dom *dom;
+
+	dom = container_of(d, struct mpam_resctrl_dom, resctrl_dom);
+	dom->mbm_local_evt_cfg = MPAM_RESTRL_EVT_CONFIG_VALID;
+	mpam_msmon_reset_all_mbwu(dom->comp);
+}
+
 static int mpam_resctrl_resource_init(struct mpam_resctrl_res *res)
 {
 	struct mpam_class *class = res->class;
@@ -970,6 +1031,7 @@ mpam_resctrl_alloc_domain(unsigned int cpu, struct mpam_resctrl_res *res)
 	dom->comp = comp;
 	INIT_LIST_HEAD(&dom->resctrl_dom.list);
 	dom->resctrl_dom.id = comp->comp_id;
+	dom->mbm_local_evt_cfg = MPAM_RESTRL_EVT_CONFIG_VALID;
 	cpumask_set_cpu(cpu, &dom->resctrl_dom.cpu_mask);
 
 	/* TODO: this list should be sorted */
-- 
2.43.0

